morf-core/src/main/java/org/alfasoftware/morf/jdbc/SqlDialect.java

```diff
@@ -3715,6 +3718,125 @@ public Collection<String> addTableFromStatements(Table table, SelectStatement se
+   * - Drops the original table
+   * - Renames the new table using the original table's name
+   * - Adds indexes from the original table
+   *
+   * @param originalTable the original table this method will replace
+   * @param selectStatement the statement used to populate the replacement table via a CTAS
+   * @return a list of statements for the operation
+   */
+  public List<String> replaceTableFromStatements(Table originalTable, SelectStatement selectStatement) {
+
+    // Due to morf's oracle table length restrictions, our temporary table name cannot be longer than 27 characters
+    final Table newTable = SchemaUtils.table("tmp_" + StringUtils.substring(originalTable.getName(), 0, 23))
+        .columns(originalTable.columns());
+
+    validateStatement(originalTable, selectStatement).ifPresent(error -> {
+      throw new IllegalArgumentException(error);
+    });
+
+    // Generate the SQL for the CTAS and post-CTAS operations
+    final List<String> createTableStatements = Lists.newArrayList();
+    createTableStatements.addAll(addTableFromStatementsWithCasting(newTable, selectStatement));
+    createTableStatements.addAll(dropTables(ImmutableList.of(originalTable), false, true));
+    createTableStatements.addAll(renameTableStatements(newTable, originalTable));
+    createTableStatements.addAll(createAllIndexStatements(originalTable));
+
+    return createTableStatements;
+  }
+
+
+  private Optional<String> validateStatement(Table table, SelectStatement selectStatement) {
+    final String separator = "\n    ";
+
+    List<String> tableColumns = table.columns().stream().map(Column::getName).collect(toList());
+    List<String> selectColumns = selectStatement.getFields().stream().map(SqlDialect::getFieldName).collect(toList());
+
+    if (tableColumns.size() != selectColumns.size()) {
+     return Optional.of("Number of table columns [" + tableColumns.size() + "] does not match number of select columns [" + selectColumns.size() + "].");
+    }
+
+    ImmutableList.Builder<Pair<String,String>> differences = ImmutableList.builder();
+    Iterator<String> tableColumnsIterator = tableColumns.iterator();
+    Iterator<String> selectColumnsIterator = selectColumns.iterator();
+
+    while (tableColumnsIterator.hasNext() && selectColumnsIterator.hasNext()) {
+      String tableColumn = tableColumnsIterator.next();
+      String selectColumn = selectColumnsIterator.next();
+      if (!tableColumn.equalsIgnoreCase(selectColumn)) {
+        differences.add(Pair.of(tableColumn, selectColumn));
+      }
+    }
+
+    List<Pair<String, String>> diffs = differences.build();
+    if (!diffs.isEmpty()) {
+      return Optional.of("Table columns do not match select columns"
+                      + "\nMismatching pairs:" + separator + diffs.stream().map(p -> p.getLeft() + " <> " + p.getRight()).collect(joining(separator)));
+    }
+
+    return Optional.empty();
+  }

```

Comment:
I would prefer the IllegalArgumentException was thrown directly from this method,
rather than returning an Optional and then reacting to it on the calling side.

Reply:
Updated to throw directly in the method